File: KthSmallestIntegerInBST.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
  /**
   * @param {TreeNode} root
   * @param {number} k
   * @return {number}
   */
  kthSmallest(root, k) {
    this.res = root.val;
    this.k = k;
    this.index = 0;

    const dfs = (node) => {
      if (!node) {
        return;
      }

      dfs(node.left);
      this.index++;
      if (this.index === this.k) {
        this.res = node.val;
        return;
      }
      dfs(node.right);
    };

    dfs(root);
    return this.res;
  }
}

File: kth_smallest_integer_in_bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        self.res = root.val
        self.index = 0
        self.k = k

        def dfs(node):
            if not node:
                return

            dfs(node.left)
            self.index += 1
            if self.index == self.k:
                self.res = node.val
                return
            dfs(node.right)

        dfs(root)
        return self.res
